File: fleet-ops/services/reel-pipeline/src/publishers/instagram.js
const DEFAULT_GRAPH_URL = 'https://graph.instagram.com';
const DEFAULT_API_VERSION = 'v22.0';
const DEFAULT_POLL_INTERVAL_MS = 3_000;
const DEFAULT_POLL_TIMEOUT_MS = 5 * 60_000;
const TERMINAL_STATUSES = new Set(['FINISHED', 'ERROR', 'EXPIRED']);
const DEFAULT_INSIGHT_METRICS = ['views', 'likes', 'comments', 'shares', 'saved'];

export class InstagramPublisher {
  constructor(options = {}) {
    this.userId = options.userId;
    this.longLivedToken = options.longLivedToken;
    this.graphUrl = (options.graphUrl ?? DEFAULT_GRAPH_URL).replace(/\/$/, '');
    this.apiVersion = options.apiVersion ?? DEFAULT_API_VERSION;
    this.pollIntervalMs = options.pollIntervalMs ?? DEFAULT_POLL_INTERVAL_MS;
    this.pollTimeoutMs = options.pollTimeoutMs ?? DEFAULT_POLL_TIMEOUT_MS;
    this.fetchImpl = options.fetchImpl ?? fetch;
    this.sleepImpl = options.sleepImpl ?? ((ms) => new Promise((resolve) => setTimeout(resolve, ms)));
    this.now = options.now ?? (() => Date.now());
  }

  base() {
    return `${this.graphUrl}/${this.apiVersion}`;
  }

  async publishReel(input) {
    if (!this.userId) throw new Error('InstagramPublisher requires userId');
    if (!this.longLivedToken) throw new Error('InstagramPublisher requires longLivedToken');
    if (!input?.videoUrl) throw new Error('publishReel requires videoUrl');
    if (!isHttpUrl(input.videoUrl)) {
      throw new Error(`InstagramPublisher.videoUrl must be a public http(s) URL (got ${input.videoUrl})`);
    }

    const containerId = await this.createContainer(input);
    const status = await this.waitForContainer(containerId);
    if (status !== 'FINISHED') {
      throw new Error(`Instagram container ${containerId} ended in ${status}`);
    }
    return this.publishContainer(containerId);
  }

  async createContainer(input) {
    const body = new URLSearchParams({
      media_type: 'REELS',
      video_url: input.videoUrl,
      caption: input.caption ?? '',
      access_token: this.longLivedToken,
    });
    if (input.shareToFeed === false) body.set('share_to_feed', 'false');
    if (input.thumbOffsetMs) body.set('thumb_offset', String(input.thumbOffsetMs));
    const res = await this.fetchImpl(`${this.base()}/${this.userId}/media`, {
      method: 'POST',
      headers: { 'content-type': 'application/x-www-form-urlencoded' },
      body: body.toString(),
    });
    if (!res.ok) {
      throw new Error(`Instagram createContainer failed ${res.status}: ${await res.text()}`);
    }
    const payload = await res.json();
    if (!payload.id) throw new Error(`Instagram createContainer missing id: ${JSON.stringify(payload)}`);
    return payload.id;
  }

  async waitForContainer(containerId) {
    const deadline = this.now() + this.pollTimeoutMs;
    let last = 'IN_PROGRESS';
    while (this.now() < deadline) {
      const url = `${this.base()}/${containerId}?fields=status_code&access_token=${encodeURIComponent(this.longLivedToken)}`;
      const res = await this.fetchImpl(url);
      if (!res.ok) throw new Error(`Instagram waitForContainer failed ${res.status}: ${await res.text()}`);
      const payload = await res.json();
      last = payload.status_code ?? 'IN_PROGRESS';
      if (TERMINAL_STATUSES.has(last)) return last;
      await this.sleepImpl(this.pollIntervalMs);
    }
    throw new Error(`Instagram container ${containerId} did not finish within ${this.pollTimeoutMs}ms (last=${last})`);
  }

  async publishContainer(containerId) {
    const body = new URLSearchParams({
      creation_id: containerId,
      access_token: this.longLivedToken,
    });
    const res = await this.fetchImpl(`${this.base()}/${this.userId}/media_publish`, {
      method: 'POST',
      headers: { 'content-type': 'application/x-www-form-urlencoded' },
      body: body.toString(),
    });
    if (!res.ok) {
      throw new Error(`Instagram media_publish failed ${res.status}: ${await res.text()}`);
    }
    const payload = await res.json();
    if (!payload.id) throw new Error(`Instagram media_publish missing id: ${JSON.stringify(payload)}`);
    return {
      mediaId: payload.id,
      url: `https://www.instagram.com/reel/${payload.id}/`,
      raw: payload,
    };
  }

  async refreshLongLivedToken() {
    if (!this.longLivedToken) throw new Error('refreshLongLivedToken requires longLivedToken');
    const url = `${this.base()}/refresh_access_token?grant_type=ig_refresh_token&access_token=${encodeURIComponent(this.longLivedToken)}`;
    const res = await this.fetchImpl(url);
    if (!res.ok) {
      throw new Error(`Instagram refresh failed ${res.status}: ${await res.text()}`);
    }
    const payload = await res.json();
    if (!payload.access_token) {
      throw new Error(`Instagram refresh missing access_token: ${JSON.stringify(payload)}`);
    }
    return {
      longLivedToken: payload.access_token,
      expiresInSeconds: Number(payload.expires_in ?? 0),
    };
  }

  async mediaInsights(mediaId, metrics = DEFAULT_INSIGHT_METRICS) {
    if (!this.longLivedToken) throw new Error('mediaInsights requires longLivedToken');
    if (!mediaId) throw new Error('mediaInsights requires mediaId');
    const metricList = metrics.join(',');
    const url = `${this.base()}/${encodeURIComponent(mediaId)}/insights?metric=${encodeURIComponent(metricList)}&access_token=${encodeURIComponent(this.longLivedToken)}`;
    const res = await this.fetchImpl(url);
    if (!res.ok) {
      throw new Error(`Instagram mediaInsights failed ${res.status}: ${await res.text()}`);
    }
    const payload = await res.json();
    return {
      provider: 'instagram',
      postId: mediaId,
      metrics: normalizeInsights(payload),
      raw: payload,
    };
  }
}

function isHttpUrl(value) {
  return typeof value === 'string' && /^https?:\/\//i.test(value);
}

function normalizeInsights(payload) {
  const metrics = {};
  for (const item of payload?.data ?? []) {
    const value = item.values?.[0]?.value ?? item.total_value?.value ?? null;
    metrics[item.name] = numberOrNull(value);
  }
  return metrics;
}

function numberOrNull(value) {
  if (value === undefined || value === null || value === '') return null;
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

File: fleet-ops/services/reel-pipeline/reel/src/publishers/instagram.rs
//! Instagram Graph reel publishing — port of `src/publishers/instagram.js`.

use std::collections::BTreeMap;
use std::thread;
use std::time::{Duration, Instant};

use anyhow::{anyhow, Context, Result};
use serde_json::Value;

const DEFAULT_GRAPH_URL: &str = "https://graph.instagram.com";
const DEFAULT_API_VERSION: &str = "v22.0";
const DEFAULT_POLL_INTERVAL_MS: u64 = 3_000;
const DEFAULT_POLL_TIMEOUT_MS: u64 = 5 * 60_000;
const DEFAULT_INSIGHT_METRICS: &[&str] = &["views", "likes", "comments", "shares", "saved"];

#[derive(Debug, Clone)]
pub struct InstagramReelInput<'a> {
    pub video_url: &'a str,
    pub caption: Option<&'a str>,
    pub share_to_feed: Option<bool>,
    pub thumb_offset_ms: Option<u64>,
}

#[derive(Debug, Clone)]
pub struct InstagramPublishResult {
    pub media_id: String,
    pub url: String,
    pub raw: Value,
}

#[derive(Debug, Clone, PartialEq)]
pub struct InstagramInsightsResult {
    pub media_id: String,
    pub metrics: BTreeMap<String, Option<f64>>,
    pub raw: Value,
}

pub struct InstagramPublisher {
    user_id: String,
    long_lived_token: String,
    graph_url: String,
    api_version: String,
    poll_interval_ms: u64,
    poll_timeout_ms: u64,
}

impl InstagramPublisher {
    pub fn from_account_fields(
        fields: &std::collections::BTreeMap<String, String>,
    ) -> Result<Self> {
        Ok(Self {
            user_id: fields
                .get("userId")
                .cloned()
                .filter(|v| !v.trim().is_empty())
                .ok_or_else(|| anyhow!("InstagramPublisher requires userId"))?,
            long_lived_token: fields
                .get("longLivedToken")
                .cloned()
                .filter(|v| !v.trim().is_empty())
                .ok_or_else(|| anyhow!("InstagramPublisher requires longLivedToken"))?,
            graph_url: fields
                .get("graphUrl")
                .cloned()
                .unwrap_or_else(|| DEFAULT_GRAPH_URL.to_string())
                .trim_end_matches('/')
                .to_string(),
            api_version: fields
                .get("apiVersion")
                .cloned()
                .unwrap_or_else(|| DEFAULT_API_VERSION.to_string()),
            poll_interval_ms: fields
                .get("pollIntervalMs")
                .and_then(|v| v.parse().ok())
                .unwrap_or(DEFAULT_POLL_INTERVAL_MS),
            poll_timeout_ms: fields
                .get("pollTimeoutMs")
                .and_then(|v| v.parse().ok())
                .unwrap_or(DEFAULT_POLL_TIMEOUT_MS),
        })
    }

    fn api_base(&self) -> String {
        format!("{}/{}", self.graph_url, self.api_version)
    }

    fn user_base(&self) -> String {
        format!("{}/{}", self.api_base(), self.user_id)
    }

    pub fn publish_reel(&self, input: &InstagramReelInput<'_>) -> Result<InstagramPublishResult> {
        if !is_http_url(input.video_url) {
            return Err(anyhow!(
                "InstagramPublisher.video_url must be a public http(s) URL (got {})",
                input.video_url
            ));
        }
        let container_id = self.create_container(input)?;
        let status = self.wait_for_container(&container_id)?;
        if status != "FINISHED" {
            return Err(anyhow!(
                "Instagram container {container_id} ended in {status}"
            ));
        }
        self.publish_container(&container_id)
    }

    pub fn media_insights(
        &self,
        media_id: &str,
        metrics: Option<&[&str]>,
    ) -> Result<InstagramInsightsResult> {
        if media_id.trim().is_empty() {
            return Err(anyhow!("media_insights requires media_id"));
        }
        let metric_list = metrics.unwrap_or(DEFAULT_INSIGHT_METRICS).join(",");
        let url = format!(
            "{}/{}/insights?metric={}&access_token={}",
            self.api_base(),
            urlencode(media_id),
            urlencode(&metric_list),
            urlencode(&self.long_lived_token),
        );
        let mut response = ureq::get(&url)
            .call()
            .map_err(|err| anyhow!("Instagram media_insights failed: {err}"))?;
        if response.status() != 200 {
            let text = response.body_mut().read_to_string().unwrap_or_default();
            return Err(anyhow!(
                "Instagram media_insights failed {}: {text}",
                response.status()
            ));
        }
        let payload: Value = serde_json::from_str(&response.body_mut().read_to_string()?)
            .context("parsing Instagram media insights response")?;
        Ok(InstagramInsightsResult {
            media_id: media_id.to_string(),
            metrics: parse_insights_metrics(&payload),
            raw: payload,
        })
    }

    fn create_container(&self, input: &InstagramReelInput<'_>) -> Result<String> {
        let mut body = format!(
            "media_type=REELS&video_url={}&caption={}&access_token={}",
            urlencode(input.video_url),
            urlencode(input.caption.unwrap_or("")),
            urlencode(&self.long_lived_token),
        );
        if input.share_to_feed == Some(false) {
            body.push_str("&share_to_feed=false");
        }
        if let Some(offset) = input.thumb_offset_ms {
            body.push_str(&format!("&thumb_offset={offset}"));
        }
        let url = format!("{}/media", self.user_base());
        let mut response = ureq::post(&url)
            .header("content-type", "application/x-www-form-urlencoded")
            .send(body.as_bytes())
            .map_err(|err| anyhow!("Instagram createContainer failed: {err}"))?;
        if response.status() != 200 {
            let text = response.body_mut().read_to_string().unwrap_or_default();
            return Err(anyhow!(
                "Instagram createContainer failed {}: {text}",
                response.status()
            ));
        }
        let payload: Value = serde_json::from_str(&response.body_mut().read_to_string()?)
            .context("parsing Instagram createContainer response")?;
        payload
            .get("id")
            .and_then(|v| v.as_str())
            .map(str::to_string)
            .ok_or_else(|| anyhow!("Instagram createContainer missing id"))
    }

    fn wait_for_container(&self, container_id: &str) -> Result<String> {
        let deadline = Instant::now() + Duration::from_millis(self.poll_timeout_ms);
        let mut last = "IN_PROGRESS".to_string();
        while Instant::now() < deadline {
            let url = format!(
                "{}/{}?fields=status_code&access_token={}",
                self.api_base(),
                urlencode(container_id),
                urlencode(&self.long_lived_token),
            );
            let mut response = ureq::get(&url)
                .call()
                .map_err(|err| anyhow!("Instagram waitForContainer failed: {err}"))?;
            if response.status() != 200 {
                let text = response.body_mut().read_to_string().unwrap_or_default();
                return Err(anyhow!(
                    "Instagram waitForContainer failed {}: {text}",
                    response.status()
                ));
            }
            let payload: Value = serde_json::from_str(&response.body_mut().read_to_string()?)?;
            last = payload
                .get("status_code")
                .and_then(|v| v.as_str())
                .unwrap_or("IN_PROGRESS")
                .to_string();
            if matches!(last.as_str(), "FINISHED" | "ERROR" | "EXPIRED") {
                return Ok(last);
            }
            thread::sleep(Duration::from_millis(self.poll_interval_ms.max(250)));
        }
        Err(anyhow!(
            "Instagram container {container_id} did not finish within {}ms (last={last})",
            self.poll_timeout_ms
        ))
    }

    fn publish_container(&self, container_id: &str) -> Result<InstagramPublishResult> {
        let body = format!(
            "creation_id={}&access_token={}",
            urlencode(container_id),
            urlencode(&self.long_lived_token),
        );
        let url = format!("{}/media_publish", self.user_base());
        let mut response = ureq::post(&url)
            .header("content-type", "application/x-www-form-urlencoded")
            .send(body.as_bytes())
            .map_err(|err| anyhow!("Instagram media_publish failed: {err}"))?;
        if response.status() != 200 {
            let text = response.body_mut().read_to_string().unwrap_or_default();
            return Err(anyhow!(
                "Instagram media_publish failed {}: {text}",
                response.status()
            ));
        }
        let payload: Value = serde_json::from_str(&response.body_mut().read_to_string()?)?;
        let media_id = payload
            .get("id")
            .and_then(|v| v.as_str())
            .ok_or_else(|| anyhow!("Instagram media_publish missing id"))?;
        Ok(InstagramPublishResult {
            media_id: media_id.to_string(),
            url: format!("https://www.instagram.com/reel/{media_id}/"),
            raw: payload,
        })
    }
}

fn parse_insights_metrics(payload: &Value) -> BTreeMap<String, Option<f64>> {
    let mut metrics = BTreeMap::new();
    if let Some(items) = payload.get("data").and_then(|v| v.as_array()) {
        for item in items {
            let Some(name) = item.get("name").and_then(|v| v.as_str()) else {
                continue;
            };
            let value = item
                .pointer("/values/0/value")
                .or_else(|| item.pointer("/total_value/value"));
            metrics.insert(name.to_string(), metric_f64(value));
        }
    }
    metrics
}

fn metric_f64(value: Option<&Value>) -> Option<f64> {
    match value? {
        Value::Number(n) => n.as_f64(),
        Value::String(s) => s.parse().ok(),
        _ => None,
    }
}

fn is_http_url(value: &str) -> bool {
    value.starts_with("http://") || value.starts_with("https://")
}

fn urlencode(input: &str) -> String {
    input
        .chars()
        .map(|c| match c {
            'A'..='Z' | 'a'..='z' | '0'..='9' | '-' | '_' | '.' | '~' => c.to_string(),
            _ => format!("%{:02X}", c as u8),
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_non_http_video_url() {
        let pub_ = InstagramPublisher {
            user_id: "1".into(),
            long_lived_token: "t".into(),
            graph_url: DEFAULT_GRAPH_URL.into(),
            api_version: DEFAULT_API_VERSION.into(),
            poll_interval_ms: 1,
            poll_timeout_ms: 1,
        };
        let err = pub_
            .publish_reel(&InstagramReelInput {
                video_url: "file:///tmp/x.mp4",
                caption: None,
                share_to_feed: None,
                thumb_offset_ms: None,
            })
            .unwrap_err();
        assert!(err.to_string().contains("http(s) URL"));
    }

    #[test]
    fn parse_insights_metrics_normalizes_values() {
        let metrics = parse_insights_metrics(&serde_json::json!({
            "data": [
                { "name": "views", "values": [{ "value": 100 }] },
                { "name": "likes", "values": [{ "value": "12" }] },
                { "name": "comments", "total_value": { "value": 3 } }
            ]
        }));
        assert_eq!(metrics.get("views"), Some(&Some(100.0)));
        assert_eq!(metrics.get("likes"), Some(&Some(12.0)));
        assert_eq!(metrics.get("comments"), Some(&Some(3.0)));
    }
}
